feat: Add rated category snaps to the Games tab

ashuntu opened the pull request:
This PR adds an ordered list of snaps based on their rating to the Games tab. Doing this required adding support for the top charts endpoints from the ratings service to the ratings client.

I've added this as a new variation of the `CategorySnapList`, so hopefully if we intend to incorporate top charts into other areas of the app, we can easily reuse the `RatedCategorySnapList`. A spinner is displayed while the list is loading; it does take a moment since we have to fetch the snap details for each snap ID.

This is more or less what it looks like, though the actual snaps displayed will not be the same:
![image](https://github.com/user-attachments/assets/01ae6053-f4b9-4efb-a644-c44b43c56bed)

UDENG-3351

anasereijo commented:
This is looking nice! 

- It would be helpful if you could share a screenshot of how it looks next to other sections, such as app cards, etc. Is this something you can share here?

- The way I designed it originally was to fit 4 cols (in the wide window) to add some interesting layout to the page. So the cards are 3 columns, and this is 4. See the screenshot below (grid in pale pink)

<img width="668" alt="Screenshot 2024-08-13 at 16 56 53" src="https://github.com/user-attachments/assets/1675deaa-ec3e-4f41-8bd2-2b0afa811e67">

- I suppose that the titles (Top rated, etc) are the same size as the ones in the Explore page, right?

- The numbers seem too big. Can you decrease the font size?


Hope this makes sense :)

ashuntu commented:
I decreased the font size of the numbers from 24 to 16. Here are some additional screenshots (with the font size change):
![image](https://github.com/user-attachments/assets/8c3b61fa-2997-4266-a5a1-e31750400c67)
![image](https://github.com/user-attachments/assets/c9aebf0c-ce95-4394-b50e-8bef45430b68)

And this is what it looks like fullscreen (I have a 1440p 16:10 display):
![image](https://github.com/user-attachments/assets/4e53d6f0-3c09-48b7-a553-4deb5b8bb0f6)

I had some trouble fitting in 4 apps in a row without compromising the titles/descriptions, but if you'd like, I can take another look at doing that.

The titles are the same size as the other page headings, yes.

anasereijo commented:
Thanks @ashuntu, the smaller numbers are looking better.

If you have the capacity, I think it is worth trying to fit the charts in 4 cols on the widest screen. Looking at the screenshot seems that you have a 'card' around the app when on hover. That isn't necessary (the whole area should be clickable though) and maybe it will save up some space to fit the 4 cols? It will also decrease the padding between the number and the icon, I think.

<img width="149" alt="Screenshot 2024-08-16 at 09 44 49" src="https://github.com/user-attachments/assets/f62afff2-0f0d-4f7b-844e-60dd703ff433">

ashuntu commented:
How does this look? (this still includes an outline on hover)

![image](https://github.com/user-attachments/assets/9742f79f-f7c3-496e-a026-f4de124fac07)

anasereijo commented:
I think this could work, thanks for trying @ashuntu!

- What is the padding between each column? It should match the padding used in the other sections.
- We don't need the outline on hover. It can be just clickable but without the outline

Thank you!

ashuntu commented:
I remove the border on hover now.

For the columns, they should have the same padding as the other cards (they use the same grid, just have different column count and aspect ratio). 

I think the numbers are slightly out of the grid in the same way that the other titles (like the F in "Featured Titles" in the screenshot you showed) are. This happens on all the pages it looks like, where the first letter in titles is slightly out of bounds. If I had to guess, that's just due to the font, but I'm not sure. Maybe someone else from the apps squad would know better than me on that.

anasereijo commented:
Thanks, @ashuntu! I can bring that up with the squad, as it's something not really related then to this PR. 

That said, LGTM! thanks for making the changes.

ashuntu commented on packages/app_center/lib/widgets/snap_grid.dart:
```diff
@@ -24,6 +25,21 @@ class AppCardGrid extends StatelessWidget {
+              (entry) => AppCard.fromRatedSnap(
+                snap: entry.value,
+                onTap: () => onTap(entry.value),
+                rating: entry.key + 1,
```
The key in this is just from the `.asMap` call, so I can get the index and give each snap a number next to it based on the index (its rank). Is there a different way you would do it?

d-loose reviewed (commented):
Could you please rebase the PR? Let me know if you need any help!

d-loose commented on packages/app_center/lib/ratings/ratings_service.dart:
```diff
@@ -32,6 +34,11 @@ class RatingsService {
 
+  Future<List<ChartData>> getChart(SnapCategoryEnum category) async {
+    await _ensureValidToken();
+    return client.getChart(Timeframe.unspecified, _jwt!, category.index);
```
The entries of `SnapCategoryEnum` and `Category` aren't identical, we'll need to to an explicit conversion between the two and can't simply use the index here

d-loose commented on packages/app_center_ratings_client/lib/src/ratings_client.dart:
```diff
@@ -53,8 +53,15 @@ class RatingsClient {
+  ]) async {
+    final request = chart_pb.GetChartRequest(
+      timeframe: timeframe.toDTO(),
+      category: category != null ? chart_pb.Category.valueOf(category) : null,
```
See previous comment, we need to map the categories explicitly between the internal representation and the grpc definition.

d-loose reviewed (changes requested):
Did some testing with the staging environment @matthew-hagemann managed to deploy now :)
There's a small bug due to mismatching category definitions, see comments for details

ashuntu commented on packages/app_center/lib/widgets/app_card.dart:
```diff
@@ -73,30 +89,58 @@ class AppCard extends StatelessWidget {
-    );
+    ];
+
+    return rating > 0
+        ? Flex(
+            direction: Axis.horizontal,
+            crossAxisAlignment: CrossAxisAlignment.start,
+            children: [
+              Center(
+                child: Text(
+                  rating.toString(),
+                  style: theme.textTheme.titleMedium,
+                ),
+              ),
+              const SizedBox(
+                width: 4,
+              ),
+              Expanded(
+                child: SmallBanner(
+                  onTap: onTap,
+                  child: Flex(
+                    direction: Axis.horizontal,
+                    children: appContent,
+                  ),
+                ),
+              ),
+            ],
+          )
+        : YaruBanner(
+            padding: const EdgeInsets.all(kCardSpacing),
+            onTap: onTap,
+            color: Theme.of(context).cardColor,
+            child: Flex(
+              direction: compact ? Axis.vertical : Axis.horizontal,
+              crossAxisAlignment: CrossAxisAlignment.start,
+              children: appContent,
+            ),
+          );
```
I was thinking of doing that originally, but the cards are used in `snap_grid` and those take types of `AppCard`. Is there a way you'd recommend splitting this up? Maybe a separate grid entirely for rated grids?

d-loose replied:
Yes, I think a separate grid would be clearer, I'd prefer some mild duplication over increasingly complex layout logic in the widgets

ashuntu commented:
@d-loose I rebased and made most of the changes you pointed out.

I added the explicit mapping for chart categories, let me know if you think there is a better way than the way I chose in https://github.com/ubuntu/app-center/pull/1740/commits/73e1b4e691606a783608b3cd2c6626c2c563049c.

d-loose commented on packages/app_center/lib/ratings/ratings_service.dart:
```diff
@@ -29,17 +30,53 @@ class RatingsService {
     await _ensureValidToken();
-    return client.getChart(Timeframe.unspecified, _jwt!, category.index);
+
+    final categoryToChart = {
```
I like doing these conversions with explicit `switch` expressions to ensure the enum is matched exhaustively, see e.g. [here](https://github.com/canonical/prompting-client/blob/02e97f3b22c339cbadd67e9977778997aee0ae30/flutter_packages/prompting_client/lib/src/prompting_client.dart#L50-L67). We'd need two separate `toProto` and `fromProto` conversions here and make sure we handle our custom UI categories properly in each of them

ashuntu replied:
I don't recall anywhere we need to convert from the ratings category to the enum category (would be `fromProto`), do you know if that conversion is actually needed anywhere?

d-loose replied:
Looks like we don't need it

d-loose commented on packages/app_center/lib/widgets/small_banner.dart:
```diff
@@ -0,0 +1,41 @@
+    final theme = Theme.of(context);
+    final borderRadius = BorderRadius.circular(kYaruBannerRadius);
+    final light = theme.brightness == Brightness.light;
+    final defaultSurfaceTintColor =
+        light ? theme.cardColor : const Color.fromARGB(255, 126, 126, 126);
```
Could we also remove the hardcoded color here? Unless absolutely necessary, it'd be great if we could avoid maintaining customizations on top of Yaru altogether.
We can probably remove the explicit `theme.cardColors` here entirely, now that https://github.com/ubuntu/app-center/pull/1818 is merged.

For reference, this is what it currently looks like:

| light | dark |
| - | - |
| ![Screenshot from 2024-10-16 13-39-55](https://github.com/user-attachments/assets/6982b856-69af-432b-ab38-3d3a79ac7d31) | ![image](https://github.com/user-attachments/assets/47d8587b-7939-46b9-9e13-d4dafd85e050) |

ashuntu replied:
Yeah, removing the `surfaceTintColor` entirely seems to retain the correct colors anyway.

d-loose replied:
A small UI nitpick I'm noticing now: the card with rank `10` is misaligned with the others, so we should make sure the rank label has a fixed width.

d-loose commented on packages/app_center/lib/games/games_page.dart:
```diff
@@ -39,8 +39,16 @@ class GamesPage extends ConsumerWidget {
-          category: SnapCategoryEnum.games,
+        const RatedCategorySnapList(
+          categories: [
+            SnapCategoryEnum.games,
+            SnapCategoryEnum.kdeGames,
+            SnapCategoryEnum.gnomeGames,
+            SnapCategoryEnum.gameLaunchers,
+            SnapCategoryEnum.gameEmulators,
+            SnapCategoryEnum.gameContentCreation,
+            SnapCategoryEnum.gameDev,
```
Since all of those map to `Category.GAMES` now, I think we only need to include `SnapCategoryEnum.games` here.

d-loose commented on packages/app_center/lib/snapd/snapd_service.dart:
```diff
@@ -1,8 +1,23 @@
+      'remote': 'true',
+      'snap-id': snapId,
+    };
+    final result =
+        await getAssertions(assertion: 'snap-declaration', params: queryParams);
+    final declaration = SnapDeclaration.fromJson(result);
+    final findResult = await find(name: declaration.snapName);
```
Sorry, another nitpick: while testing locally I noticed this can throw an exception if the snap id can't be resolved (they're randomly generated in the tests, so almost always invalid). That usually shouldn't be an issue in production, but doesn't hurt to add a check here :)

d-loose reviewed (commented):
Thanks, those are great improvements! I have two more tiny nitpicks and would be happy to merge this soon :)

@sminez, @matthew-hagemann should we wait until the ratings service with support for charts is deployed before merging this? Otherwise we'd still need to add some UI fallback state that hides the top charts in case the service doesn't support the feature.

ashuntu commented:
> Otherwise we'd still need to add some UI fallback state that hides the top charts in case the service doesn't support the feature.

I added fallback to the old behavior (just a regular `CategorySnapList`) when the ratings service for charts isn't available https://github.com/ubuntu/app-center/pull/1740/commits/23375de704a8d3d5e434c660085b8a2c09643eb2.

d-loose reviewed (approved):
Awesome, thanks! I'd still wait until tomorrow morning (European time) to merge this, if that's alright :)

ashuntu commented:
I added a try/catch here instead for the assertions endpoint to handle the YAML errors we talked about. So for now the rated grids will just ignore any snap with invalid YAML assertions until a raw response is sorted in snapd.dart.

d-loose commented:
@sminez I've pushed a `gaming-top-charts` branch (currently the same as `main`) and changed the base of this PR to target that one, so we can land additional changes before merging to `main` without bothering @ashuntu too much :)

sminez commented:
@d-loose thanks for re-pointing the PR to the new branch :ok_hand: please merge at your leisure :sunglasses:
